File: LeetCode/13_Roman_to_Integer/13_Roman_to_Integer.go
package main

import "fmt"

func romanToInt(s string) int {
	symbolMap := map[rune]int{
		'I': 1,
		'V': 5,
		'X': 10,
		'L': 50,
		'C': 100,
		'D': 500,
		'M': 1000,
	}
	prev := rune(0)
	result := 0
	for _, c := range s {
		result += symbolMap[c]
		if (c == 'V' || c == 'X') && prev == 'I' {
			result -= 2 * symbolMap[prev]
		} else if (c == 'L' || c == 'C') && prev == 'X' {
			result -= 2 * symbolMap[prev]
		} else if (c == 'D' || c == 'M') && prev == 'C' {
			result -= 2 * symbolMap[prev]
		}
		prev = c
	}
	return result
}

func main() {
	s := "MCMXCIV"
	fmt.Println(romanToInt(s))
}

File: LeetCode/13_Roman_to_Integer/13_Roman_to_Integer.kt
class Solution {
    fun romanToInt(s: String): Int {
        val symbolMap = mapOf(
            'I' to 1,
            'V' to 5,
            'X' to 10,
            'L' to 50,
            'C' to 100,
            'D' to 500,
            'M' to 1000
        )
        // var prev = 0.toChar()
        var prev = '\u0000'
        var result = 0
        for (c in s) {
            result += symbolMap[c]!!
            if ((c == 'V' || c == 'X') && prev == 'I') {
                result -= 2 * symbolMap[prev]!!
            } else if ((c == 'L' || c == 'C') && prev == 'X') {
                result -= 2 * symbolMap[prev]!!
            } else if ((c == 'D' || c == 'M') && prev == 'C') {
                result -= 2 * symbolMap[prev]!!
            }
            prev = c
        }
        return result
    }
}

fun main() {
    val solution = Solution()
    val s = "MCMXCIV"
    print(solution.romanToInt(s))
}